webapp/tests/karma/ts/services/startup-modals.service.spec.ts

```diff
@@ -0,0 +1,135 @@
+import { TestBed } from '@angular/core/testing';
+import sinon from 'sinon';
+import { expect } from 'chai';
+
+import {SessionService} from '@mm-services/session.service';
+import {SettingsService} from '@mm-services/settings.service';
+import {StartupModalsService} from '@mm-services/startup-modals.service';
+import {UpdateUserService} from '@mm-services/update-user.service';
+import {UserSettingsService} from '@mm-services/user-settings.service';
+import {ModalService} from '@mm-modals/mm-modal/mm-modal';
+import {TourService} from "@mm-services/tour.service";
+
+
+describe('StartupModalsService', () => {
+
+  let service: StartupModalsService;
+  let updateUserService;
+  let sessionService;
+  let modalService;
+  let tourService;
+
+  afterEach(() => {
+    sinon.restore();
+  });
+
+  function getProviders(toursResult) {
```

Comment:
Yes the problem is that I in both tests the configuration is almost the same, except for the one bean that in its constructor executes a call to another service that is mocked, and the execution happens only one time at the beginning of the initialization, so I can refactor like you said to make 2 `beforeEach` functions, but each one will only be used by one test case.

Reply:
I think you can change the moked result before you inject the service.

Reply:
I tried what you said in this [fork](https://github.com/medic/cht-core/blob/4750-angular-10-tour-startup-modals-spec-fix/webapp/tests/karma/ts/services/startup-modals.service.spec.ts) of the branch but the stub does not work, I got:

```
      ✖ "before each" hook for "Tour modal should not be displayed if no tours are available"
        Chrome Headless 86.0.4240.193 (Linux x86_64)
      TypeError: Cannot read property 'then' of undefined
          at new StartupModalsService (_karma_webpack_/src/ts/services/startup-modals.service.ts:71:32)
```
Like the override of the stub in the method itself is executed later of the service initialization.

I can try a double nested `describe` section, with a main `beforeEach` and a `beforeEach` on each nested `describe`, but not sure if it will work and still will only have one test on each describe/describe block.